File: backend/go/internal/server/groups.go
package server

// Voice groups (spec 6.12): up to 15 sessions that hear each other non-positionally,
// across worlds and Minecraft servers. State is guarded by s.mu.

import (
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"math/big"
	"sort"
	"strings"
	"time"

	"github.com/RavoxX/MCVoice/backend/go/internal/protocol"
)

// outOfWorldGrace: out of a world for longer ends the membership (respawn, dimension or
// server switch are shorter).
const outOfWorldGrace = 10 * time.Second

const passwordFailsPerMin = 5

type group struct {
	id      string
	salt    []byte
	pwHash  []byte // SHA-256(salt || password); nil = open. Never logged, never sent.
	members []*Session
}

type outMsg struct {
	sess *Session
	msg  []byte
}

type memberJSON struct {
	UUID string `json:"uuid"`
	Name string `json:"name"`
}

func pwHash(salt []byte, pw string) []byte {
	h := sha256.New()
	h.Write(salt)
	h.Write([]byte(pw))
	return h.Sum(nil)
}

func (g *group) passwordOK(given *string) bool {
	if g.pwHash == nil {
		return true
	}
	if given == nil {
		return false
	}
	return subtle.ConstantTimeCompare(pwHash(g.salt, *given), g.pwHash) == 1
}

func (s *Server) newGroupIDLocked() string {
	n := big.NewInt(int64(len(protocol.GroupIDAlphabet)))
	for {
		b := make([]byte, protocol.GroupIDLen)
		for i := range b {
			k, err := rand.Int(rand.Reader, n)
			if err != nil {
				panic(err)
			}
			b[i] = protocol.GroupIDAlphabet[k.Int64()]
		}
		if _, taken := s.groups[string(b)]; !taken {
			return string(b)
		}
	}
}

func membersJSON(g *group) []memberJSON {
	out := make([]memberJSON, 0, len(g.members))
	for _, m := range g.members {
		out = append(out, memberJSON{UUID: m.ident.UUID, Name: m.ident.Username})
	}
	return out
}

func (s *Server) queueLocked(sess *Session, msg []byte) {
	s.outbox = append(s.outbox, outMsg{sess, msg})
}

func (s *Server) announceLocked(g *group, except *Session) {
	msg := marshal(map[string]any{"type": "group_update", "id": g.id, "members": membersJSON(g)})
	for _, m := range g.members {
		if m != except {
			s.queueLocked(m, msg)
		}
	}
}

// leaveGroupLocked removes sess from its group; reason ("" = none) is sent to the leaver.
func (s *Server) leaveGroupLocked(sess *Session, reason string) {
	id := sess.peer.Group
	if id == "" {
		return
	}
	sess.peer.Group = ""
	if reason != "" {
		s.queueLocked(sess, marshal(map[string]any{"type": "group_left", "id": id, "reason": reason}))
	}
	g := s.groups[id]
	if g == nil {
		return
	}
	for i, m := range g.members {
		if m == sess {
			g.members = append(g.members[:i], g.members[i+1:]...)
			break
		}
	}
	if len(g.members) == 0 {
		delete(s.groups, id)
	} else {
		s.announceLocked(g, nil)
	}
}

func (s *Server) joinGroupLocked(sess *Session, g *group) {
	sess.peer.Group = g.id
	found := false
	for _, m := range g.members {
		if m == sess {
			found = true
		}
	}
	if !found {
		g.members = append(g.members, sess)
	}
	s.queueLocked(sess, marshal(map[string]any{"type": "group_joined", "group": map[string]any{
		"id": g.id, "max": protocol.GroupMaxMembers, "password": g.pwHash != nil, "members": membersJSON(g)}}))
	s.announceLocked(g, sess)
}

// flushLocked releases s.mu and sends the queued group messages.
func (s *Server) unlockAndFlush() {
	out := s.outbox
	s.outbox = nil
	s.mu.Unlock()
	for _, o := range out {
		o.sess.send(o.msg)
	}
}

func (s *Server) groupsAllowed(sess *Session) bool {
	if !sess.groupsCap {
		sess.send(errorFrame(protocol.CodeUnknownMessage, "groups capability not negotiated", false))
	}
	return sess.groupsCap
}

func (s *Server) groupList(sess *Session, query *string) {
	if !s.groupsAllowed(sess) {
		return
	}
	type entry struct {
		ID       string `json:"id"`
		Members  int    `json:"members"`
		Max      int    `json:"max"`
		Password bool   `json:"password"`
	}
	s.mu.RLock()
	l := make([]entry, 0, len(s.groups))
	q := ""
	if query != nil {
		q = strings.ToUpper(*query)
	}
	for _, g := range s.groups {
		if q != "" && !strings.Contains(g.id, q) {
			continue
		}
		l = append(l, entry{g.id, len(g.members), protocol.GroupMaxMembers, g.pwHash != nil})
	}
	s.mu.RUnlock()
	sort.Slice(l, func(i, j int) bool {
		if l[i].Members != l[j].Members {
			return l[i].Members > l[j].Members
		}
		return l[i].ID < l[j].ID
	})
	if len(l) > protocol.GroupListMax {
		l = l[:protocol.GroupListMax]
	}
	sess.send(marshal(map[string]any{"type": "group_list", "groups": l}))
}

func (s *Server) groupCreate(sess *Session, m *protocol.GroupCreate) {
	if !s.groupsAllowed(sess) {
		return
	}
	s.mu.Lock()
	if !sess.peer.InWorld {
		s.mu.Unlock()
		sess.send(errorFrame(protocol.CodeNotInWorld, "join a server first", false))
		return
	}
	s.leaveGroupLocked(sess, "replaced")
	g := &group{id: s.newGroupIDLocked()}
	if m.Password != nil {
		g.salt = randomBytes(16)
		g.pwHash = pwHash(g.salt, *m.Password)
	}
	s.groups[g.id] = g
	s.joinGroupLocked(sess, g)
	s.unlockAndFlush()
}

func (s *Server) groupJoin(sess *Session, m *protocol.GroupJoin, now time.Time) {
	if !s.groupsAllowed(sess) {
		return
	}
	id := strings.ToUpper(*m.ID)
	s.mu.Lock()
	fail := func(code, msg string) {
		s.mu.Unlock()
		sess.send(errorFrame(code, msg, false))
	}
	if !sess.peer.InWorld {
		fail(protocol.CodeNotInWorld, "join a server first")
		return
	}
	if now.Sub(sess.groupFailWindow) > time.Minute {
		sess.groupFailWindow, sess.groupFails = now, 0
	}
	if sess.groupFails >= passwordFailsPerMin {
		fail(protocol.CodeRateLimited, "too many wrong group passwords")
		return
	}
	g := s.groups[id]
	already := sess.peer.Group == id
	switch {
	case g == nil:
		fail(protocol.CodeGroupNotFound, "no such group")
		return
	case already:
	case len(g.members) >= protocol.GroupMaxMembers:
		fail(protocol.CodeGroupFull, "group is full")
		return
	case !g.passwordOK(m.Password):
		sess.groupFails++
		fail(protocol.CodeGroupPassword, "wrong or missing password")
		return
	}
	if !already {
		s.leaveGroupLocked(sess, "replaced")
	}
	s.joinGroupLocked(sess, g)
	s.unlockAndFlush()
}

func (s *Server) groupLeave(sess *Session) {
	if !s.groupsAllowed(sess) {
		return
	}
	s.mu.Lock()
	s.leaveGroupLocked(sess, "left")
	s.unlockAndFlush()
}

// groupTick ends memberships of sessions out of a world longer than the grace period.
func (s *Server) groupTick(now time.Time) {
	s.mu.Lock()
	for _, sess := range s.byConn {
		if sess.peer.Group != "" && !sess.outOfWorldSince.IsZero() && now.Sub(sess.outOfWorldSince) > outOfWorldGrace {
			s.leaveGroupLocked(sess, "not_in_world")
		}
	}
	s.unlockAndFlush()
}

File: backend/rust/src/server/groups.rs
//! Voice groups (spec 6.12): up to 15 sessions that hear each other non-positionally,
//! across worlds and Minecraft servers. State lives in the hub, under its lock.

use std::sync::Arc;
use std::time::{Duration, Instant};

use rand::Rng;
use serde_json::{json, Value};
use sha2::{Digest, Sha256};
use subtle::ConstantTimeEq;

use super::session::Session;
use super::{Hub, Server};
use crate::protocol::control::{
    code, error_frame, GroupCreate, GroupJoin, GROUP_ID_ALPHABET, GROUP_ID_LEN, GROUP_LIST_MAX, GROUP_MAX_MEMBERS,
};

/// Out of a world for longer than this ends the membership (respawn, dimension or server switch are shorter).
pub(crate) const OUT_OF_WORLD_GRACE: Duration = Duration::from_secs(10);
const PASSWORD_FAILS_PER_MIN: u32 = 5;

pub(crate) struct Group {
    pub id: String,
    /// salt, SHA-256(salt || password); never logged, never sent
    password: Option<([u8; 16], [u8; 32])>,
    /// connection ids in join order
    pub members: Vec<u64>,
}

fn hash(salt: &[u8; 16], password: &str) -> [u8; 32] {
    let mut h = Sha256::new();
    h.update(salt);
    h.update(password.as_bytes());
    h.finalize().into()
}

impl Group {
    fn new(id: String, password: Option<&str>) -> Group {
        let password = password.map(|p| {
            let salt: [u8; 16] = rand::thread_rng().gen();
            (salt, hash(&salt, p))
        });
        Group {
            id,
            password,
            members: Vec::new(),
        }
    }

    fn password_ok(&self, given: Option<&str>) -> bool {
        match (&self.password, given) {
            (None, _) => true,
            (Some((salt, want)), Some(p)) => hash(salt, p).ct_eq(want).into(),
            (Some(_), None) => false,
        }
    }
}

impl Hub {
    fn new_group_id(&self) -> String {
        let mut rng = rand::thread_rng();
        loop {
            let id: String = (0..GROUP_ID_LEN)
                .map(|_| GROUP_ID_ALPHABET[rng.gen_range(0..GROUP_ID_ALPHABET.len())] as char)
                .collect();
            if !self.groups.contains_key(&id) {
                return id;
            }
        }
    }

    fn members_json(&self, g: &Group) -> Vec<Value> {
        g.members
            .iter()
            .filter_map(|c| self.by_conn.get(c))
            .map(|s| json!({"uuid": s.ident.uuid, "name": s.ident.username}))
            .collect()
    }

    fn queue(&mut self, conn: u64, msg: String) {
        if let Some(s) = self.by_conn.get(&conn) {
            self.outbox.push((s.clone(), msg));
        }
    }

    fn announce_members(&mut self, id: &str, except: Option<u64>) {
        let Some(g) = self.groups.get(id) else { return };
        let msg = json!({"type": "group_update", "id": g.id, "members": self.members_json(g)}).to_string();
        let targets: Vec<u64> = g.members.iter().copied().filter(|c| Some(*c) != except).collect();
        for c in targets {
            self.queue(c, msg.clone());
        }
    }

    /// Remove `conn` from its group (if any). `reason` is sent to the leaver unless None.
    pub(crate) fn leave_group(&mut self, conn: u64, reason: Option<&str>) {
        let Some(ps) = self.peers.get_mut(&conn) else { return };
        let id = std::mem::take(&mut ps.peer.group);
        if id.is_empty() {
            return;
        }
        if let Some(r) = reason {
            let msg = json!({"type": "group_left", "id": id, "reason": r}).to_string();
            self.queue(conn, msg);
        }
        let empty = match self.groups.get_mut(&id) {
            Some(g) => {
                g.members.retain(|c| *c != conn);
                g.members.is_empty()
            }
            None => false,
        };
        if empty {
            self.groups.remove(&id);
        } else {
            self.announce_members(&id, None);
        }
    }

    fn join_group(&mut self, conn: u64, id: &str) {
        if let Some(ps) = self.peers.get_mut(&conn) {
            ps.peer.group = id.to_string();
        }
        let Some(g) = self.groups.get_mut(id) else { return };
        if !g.members.contains(&conn) {
            g.members.push(conn);
        }
        let g = &self.groups[id];
        let joined = json!({"type": "group_joined", "group": {
            "id": g.id, "max": GROUP_MAX_MEMBERS, "password": g.password.is_some(), "members": self.members_json(g)
        }})
        .to_string();
        self.queue(conn, joined);
        self.announce_members(id, Some(conn));
    }

    fn group_list_json(&self, query: Option<&str>) -> String {
        let q = query.map(|q| q.to_ascii_uppercase());
        let mut l: Vec<&Group> = self
            .groups
            .values()
            .filter(|g| q.as_deref().map_or(true, |q| g.id.contains(q)))
            .collect();
        l.sort_by(|a, b| b.members.len().cmp(&a.members.len()).then_with(|| a.id.cmp(&b.id)));
        let groups: Vec<Value> = l
            .into_iter()
            .take(GROUP_LIST_MAX)
            .map(|g| json!({"id": g.id, "members": g.members.len(), "max": GROUP_MAX_MEMBERS, "password": g.password.is_some()}))
            .collect();
        json!({"type": "group_list", "groups": groups}).to_string()
    }

    /// Messages queued by group operations, to send once the hub lock is released.
    pub(crate) fn take_outbox(&mut self) -> Vec<(Arc<Session>, String)> {
        std::mem::take(&mut self.outbox)
    }
}

pub(crate) fn send_all(out: Vec<(Arc<Session>, String)>) {
    for (s, m) in out {
        s.send(m);
    }
}

impl Server {
    fn group_session_ok(&self, sess: &Session, hub: &Hub) -> bool {
        let caps = hub.peers.get(&sess.conn_id).is_some_and(|p| p.groups_cap);
        if !caps {
            sess.send(error_frame(code::UNKNOWN_MESSAGE, "groups capability not negotiated", false));
        }
        caps
    }

    pub(crate) fn group_list(&self, sess: &Session, query: Option<&str>) {
        let hub = self.hub.read().unwrap();
        if !self.group_session_ok(sess, &hub) {
            return;
        }
        let msg = hub.group_list_json(query);
        drop(hub);
        sess.send(msg);
    }

    pub(crate) fn group_create(&self, sess: &Session, m: GroupCreate) {
        let mut hub = self.hub.write().unwrap();
        if !self.group_session_ok(sess, &hub) {
            return;
        }
        if !hub.peers.get(&sess.conn_id).is_some_and(|p| p.peer.in_world) {
            drop(hub);
            sess.send(error_frame(code::NOT_IN_WORLD, "join a server first", false));
            return;
        }
        hub.leave_group(sess.conn_id, Some("replaced"));
        let id = hub.new_group_id();
        hub.groups.insert(id.clone(), Group::new(id.clone(), m.password.as_deref()));
        hub.join_group(sess.conn_id, &id);
        let out = hub.take_outbox();
        drop(hub);
        send_all(out);
    }

    pub(crate) fn group_join(&self, sess: &Session, m: GroupJoin, now: Instant) {
        let id = m.id.unwrap_or_default().to_ascii_uppercase();
        let mut hub = self.hub.write().unwrap();
        if !self.group_session_ok(sess, &hub) {
            return;
        }
        let Some(ps) = hub.peers.get_mut(&sess.conn_id) else { return };
        if !ps.peer.in_world {
            drop(hub);
            sess.send(error_frame(code::NOT_IN_WORLD, "join a server first", false));
            return;
        }
        if now.duration_since(ps.group_fail_window) > Duration::from_secs(60) {
            ps.group_fail_window = now;
            ps.group_fails = 0;
        }
        if ps.group_fails >= PASSWORD_FAILS_PER_MIN {
            drop(hub);
            sess.send(error_frame(code::RATE_LIMITED, "too many wrong group passwords", false));
            return;
        }
        let already = ps.peer.group == id;
        let err = match hub.groups.get(&id) {
            None => Some((code::GROUP_NOT_FOUND, "no such group")),
            Some(_) if already => None,
            Some(g) if g.members.len() >= GROUP_MAX_MEMBERS => Some((code::GROUP_FULL, "group is full")),
            Some(g) if !g.password_ok(m.password.as_deref()) => Some((code::GROUP_PASSWORD, "wrong or missing password")),
            Some(_) => None,
        };
        if let Some((c, msg)) = err {
            if c == code::GROUP_PASSWORD {
                if let Some(ps) = hub.peers.get_mut(&sess.conn_id) {
                    ps.group_fails += 1;
                }
            }
            drop(hub);
            sess.send(error_frame(c, msg, false));
            return;
        }
        if !already {
            hub.leave_group(sess.conn_id, Some("replaced"));
        }
        hub.join_group(sess.conn_id, &id);
        let out = hub.take_outbox();
        drop(hub);
        send_all(out);
    }

    pub(crate) fn group_leave(&self, sess: &Session) {
        let mut hub = self.hub.write().unwrap();
        if !self.group_session_ok(sess, &hub) {
            return;
        }
        hub.leave_group(sess.conn_id, Some("left"));
        let out = hub.take_outbox();
        drop(hub);
        send_all(out);
    }

    /// End memberships of sessions that have been out of a world longer than the grace period.
    pub(crate) fn group_tick(&self, now: Instant) {
        let mut hub = self.hub.write().unwrap();
        let expired: Vec<u64> = hub
            .peers
            .iter()
            .filter(|(_, p)| !p.peer.group.is_empty() && p.out_of_world_since.is_some_and(|t| now.duration_since(t) > OUT_OF_WORLD_GRACE))
            .map(|(c, _)| *c)
            .collect();
        for c in expired {
            hub.leave_group(c, Some("not_in_world"));
        }
        let out = hub.take_outbox();
        drop(hub);
        send_all(out);
    }
}
